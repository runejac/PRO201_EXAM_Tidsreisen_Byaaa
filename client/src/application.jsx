import { BrowserRouter, Route, Routes } from "react-router-dom";
import { LoginPage } from "./pages/LoginPage.jsx";
import { MapPage } from "./pages/MapPage.jsx";
import { Quiz } from "./pages/capsules/quizCapsule/Quiz.jsx";
import { History } from "./pages/capsules/HistoryCapsule/History.jsx";
import { MyFindings } from "./pages/MyFindings.jsx";
import { UserContext } from "./contexts/userContext.jsx";
import { Profile } from "./pages/Profile.jsx";
import { NotFound } from "./components/NotFound.jsx";
import { BackButton } from "./components/BackButton";
import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useLoading } from "./helpers/useLoading.jsx";
import WithoutNavbar from "./components/WithoutNavbar";
import WithNavbar from "./components/WithNavbar";
import Audio from "./pages/capsules/AudioCapsule/Audio.jsx";
import Camera from "./components/Camera";
import WithBackButton from "./components/WithBackButton.jsx";

export const User = createContext("");
export const MapContext = React.createContext(null);

export function Application() {
  const [user, setUser] = useState([]);
  const [map, setMap] = useState(null);

  const { getUser } = useContext(UserContext);
  const { data: username, reload, loading, error } = useLoading(getUser);

  const providerValue = useMemo(() => ({ user, setUser }), [user, setUser]);
  const mapProvider = useMemo(() => ({ map, setMap }), [map, setMap]);

  useEffect(() => {
    username ? setUser(...username, username) : reload();
  }, [username]);

  return (
    <User.Provider value={providerValue}>
      <MapContext.Provider value={mapProvider}>
        <BrowserRouter>
          <main>
            <Routes>
              <Route element={<WithoutNavbar />}>
                <Route path="/" element={<LoginPage />} />
              </Route>
              <Route element={<WithNavbar />}>
                <Route path="/map" element={<MapPage />} />
                <Route path="/camera" element={<Camera />} />
                <Route element={<WithBackButton />}>
                  <Route path="/quiz/:id" element={<Quiz />} />
                  <Route path="/history/:id" element={<History />} />
                  <Route
                    path="/myfindings"
                    element={<MyFindings loading={loading} error={error} />}
                  />
                  <Route path="/audio/:id" element={<Audio />} />
                  <Route path="/profile" element={<Profile />} />
                  <Route path={"*"} element={<NotFound />} />
                </Route>
              </Route>
            </Routes>
          </main>
        </BrowserRouter>
      </MapContext.Provider>
    </User.Provider>
  );
}
