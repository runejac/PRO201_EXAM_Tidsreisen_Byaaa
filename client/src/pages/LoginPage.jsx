import "../css/login.css";
import logo from "../assets/images/rlogo.svg";
import { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { CapsuleButtonYellow } from "../components/CapsuleButton.jsx";
import { checkUser, postJSON } from "../helpers/http.jsx";
import Alert from "react-bootstrap/Alert";
import { MapContext, User } from "../application.jsx";

export function LoginPage() {
  const { user, setUser } = useContext(User);

  const [newUser, setNewUser] = useState("");

  const [exists, setExists] = useState("");
  const [oldUser, setOldUser] = useState("");
  const { setMap, map } = useContext(MapContext);

  const navigate = useNavigate();

  useEffect(() => {
    map ? setMap(null) : "";
  }, [map, setMap]);

  async function handleSubmit(event) {
    event.preventDefault();

    const { name, intro, walk, points, finishedCapsules } = user;

    if (event.nativeEvent.submitter.value !== "") {
      await postJSON("api/login", { user: oldUser.name, force: true });
      setUser({
        name: oldUser.name,
        intro: oldUser.intro,
        walk: oldUser.walk,
        points: oldUser.points,
        finishedCapsules: oldUser.finishedCapsules,
      });
      navigate("/map");
    } else {
      const res = await checkUser(`name=${newUser}`);

      if (res.length > 0) {
        setExists(true);
        setOldUser(res[0]);
      } else {
        console.log("creating user: " + newUser);
        await postJSON("/api/login", { user: newUser });
        setUser({
          name: newUser,
          intro: true,
          walk: false,
          points: 0,
          finishedCapsules: [],
        });
        navigate("/map");
      }
    }
  }

  return (
    <section id="login">
      <div id="logo">
        <h1 className="logo">Tidsreisen</h1>
      </div>
      <form onSubmit={handleSubmit}>
        <div className="input-div">
          <input
            data-testid={"input-login"}
            placeholder="Brukernavn"
            type="text"
            name="username"
            required
            value={newUser}
            onChange={(e) => {
              setNewUser(e.target.value);
              setExists(false);
            }}
          />
        </div>
        {exists ? (
          <>
            <Alert variant="danger">
              <Alert.Heading>
                Brukernavnet "{oldUser.name}" eksisterer allerede.
              </Alert.Heading>
              <p>Er dette deg?</p>
            </Alert>
            <CapsuleButtonYellow
              submit={"submit"}
              buttonText={"Ja, gå videre"}
              exists={oldUser.name}
            />
          </>
        ) : (
          <div>
            <CapsuleButtonYellow submit={"submit"} buttonText={"Gå videre"} />
          </div>
        )}
      </form>
      <div id={"logo-container"}>
        <img id="logoPic" width={100} src={logo} alt="Rælingen logo" />
      </div>
    </section>
  );
}
