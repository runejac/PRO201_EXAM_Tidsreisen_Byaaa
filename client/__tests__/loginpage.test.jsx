/**
 * @jest-environment jsdom
 */

import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter } from "react-router-dom";
import { LoginPage } from "../src/pages/LoginPage.jsx";
import renderer from "react-test-renderer";
import { MapContext } from "../src/application.jsx";
import { act, Simulate } from "react-dom/test-utils";
import { fireEvent } from "@testing-library/react";

//TODO: Simulere at man skriver i input-felt og trykker på knapp,
//      både med ny bruker og eksisterende

describe("LoginPage", () => {
  it("shows snapshot", async () => {
    const context = jest.fn();
    const component = renderer.create(
      <MemoryRouter>
        <MapContext.Provider value={{ context }}>
          <LoginPage />
        </MapContext.Provider>
      </MemoryRouter>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should test input on LoginPage", async function () {
    const context = jest.fn();
    const container = document.createElement("div");

    ReactDOM.render(
      <MemoryRouter>
        <MapContext.Provider value={{ context }}>
          <LoginPage />
        </MapContext.Provider>
      </MemoryRouter>,
      container
    );

    const inputTag = container.querySelector("[data-testid=input-login]");
    fireEvent.change(inputTag, { target: { value: "Hildur" } });
    expect(inputTag.value).toBe("Hildur");
  });
});
