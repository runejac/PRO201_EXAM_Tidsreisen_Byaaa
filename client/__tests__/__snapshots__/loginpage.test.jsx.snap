// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage shows snapshot 1`] = `
<section
  id="login"
>
  <div
    id="logo"
  >
    <h1
      className="logo"
    >
      Tidsreisen
    </h1>
  </div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="input-div"
    >
      <input
        data-testid="input-login"
        name="username"
        onChange={[Function]}
        placeholder="Brukernavn"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div>
      <button
        id="button-capsules-yellow"
        type="submit"
      >
        Gå videre
      </button>
    </div>
  </form>
  <div
    id="logo-container"
  >
    <img
      alt="Rælingen logo"
      id="logoPic"
      src="test-file-stub"
      width={100}
    />
  </div>
</section>
`;
